lib/recurly/apple-pay/apple-pay.js

```diff
@@ -383,6 +394,21 @@ export class ApplePay extends Emitter {
   }
 }
 
+function registerEventHandler (emitter, eventName, handler) {
+  return function (event) {
+    debug(eventName, event);
+
+    // listener accepts the `done` callback as an argument
+    if (!emitter.listeners(eventName).some(l => l.length >= 2)) {
+      emitter.emit(eventName, event);
+      handler(event);
+      return;
+    }
+
+    emitter.emit(eventName, event, (...args) => handler(event, ...args));
+  };
+}
```

Comment:
I think that this can become brittle when the integrator listens to the event more than once, where at least two of the listeners include a callback:

```
applePay.on('paymentAuthorized', ({ payment: { shippingContact } }, complete) => {
  if (shippingContact === "test@example.com") complete({ errors: [new window.ApplePayError(...)] });
  else complete();
});

applePay.on('paymentAuthorized', (event, complete) => {
  doSomethingAsync(event).then(() => complete());
});
```
This can introduce a race condition where the second listener would (a) be abandoned and (b) call the handler again, perhaps causing the ApplePay instance to produce an error?

There isn't a clean way to support multiple listeners with various return states, since it wouldn't be proper to collate or merge the return values...

So the alternative would be to prevent the integrator from listening with a callback more than once.

```
on (eventName, newListener) {
  if (this.listeners(eventName).some(l => l.length >= 2) && newListener.length >= 2) {
    // definitely open to suggestions on wording of this error.
    throw new Error(`The event '${eventName}' may only have one registered callback listener`);
  }

  return super.on(eventName, newListener);
}
```

Reply:
updated to support the explicit `options.callbacks`